File: python/examples/g24.py
# To display optimization information (none by default)
import argparse
import logging

import numpy as np

import egobox as egx

logging.basicConfig(level=logging.INFO)

xspecs_g24 = [[0.0, 3.0], [0.0, 4.0]]
n_cstr_g24 = 2


def parse_args():
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(description="G24 constrained optimization example")
    parser.add_argument(
        "--infill-opt",
        type=str,
        default="COBYLA",
        choices=["COBYLA", "SLSQP", "IPOPT"],
        help="Infill optimizer to use (default: COBYLA)",
    )
    return parser.parse_args()


# Objective
def G24(point):
    """
    Function g24
    1 global optimum y_opt = -5.5080 at x_opt =(2.3295, 3.1785)
    """
    p = np.atleast_2d(point)
    return -p[:, 0] - p[:, 1]


# Constraints < 0
def G24_c1(point):
    p = np.atleast_2d(point)
    return (
        -2.0 * p[:, 0] ** 4.0
        + 8.0 * p[:, 0] ** 3.0
        - 8.0 * p[:, 0] ** 2.0
        + p[:, 1]
        - 2.0
    )


def G24_c2(point):
    p = np.atleast_2d(point)
    return (
        -4.0 * p[:, 0] ** 4.0
        + 32.0 * p[:, 0] ** 3.0
        - 88.0 * p[:, 0] ** 2.0
        + 96.0 * p[:, 0]
        + p[:, 1]
        - 36.0
    )


# Grouped evaluation
def g24(point):
    p = np.atleast_2d(point)
    return np.array([G24(p), G24_c1(p), G24_c2(p)]).T


def main():
    args = parse_args()

    # Map string argument to InfillOptimizer enum
    infill_optimizer_map = {
        "COBYLA": egx.InfillOptimizer.COBYLA,
        "SLSQP": egx.InfillOptimizer.SLSQP,
        "IPOPT": egx.InfillOptimizer.IPOPT,
    }
    infill_optimizer = infill_optimizer_map[args.infill_opt]

    # Configure the optimizer. See help(egor) for options
    egor = egx.Egor(
        xspecs_g24,
        n_doe=10,
        n_cstr=n_cstr_g24,
        cstr_tol=[1e-3] * n_cstr_g24,
        infill_strategy=egx.InfillStrategy.WB2,
        infill_optimizer=infill_optimizer,
        target=-5.50,  # known reference objective value
    )

    optim = egor.minimize(g24, max_iters=30)
    print(
        f"Optimization f={optim.result.y_opt} at {optim.result.x_opt} "
        f"(using {args.infill_opt} infill optimizer)"
    )


if __name__ == "__main__":
    main()

File: crates/ego/examples/g24.rs
use clap::Parser;
use egobox_doe::{Lhs, SamplingMethod};
use egobox_ego::{EgorBuilder, InfillOptimizer};
use ndarray::{Array2, ArrayBase, ArrayView2, Data, Ix1, Zip, array};

// Objective
fn g24(x: &ArrayBase<impl Data<Elem = f64>, Ix1>) -> f64 {
    // Function G24: 1 global optimum y_opt = -5.5080 at x_opt =(2.3295, 3.1785)
    -x[0] - x[1]
}

// Constraints < 0
fn g24_c1(x: &ArrayBase<impl Data<Elem = f64>, Ix1>) -> f64 {
    -2.0 * x[0].powf(4.0) + 8.0 * x[0].powf(3.0) - 8.0 * x[0].powf(2.0) + x[1] - 2.0
}

fn g24_c2(x: &ArrayBase<impl Data<Elem = f64>, Ix1>) -> f64 {
    -4.0 * x[0].powf(4.0) + 32.0 * x[0].powf(3.0) - 88.0 * x[0].powf(2.0) + 96.0 * x[0] + x[1]
        - 36.0
}

fn f_g24(x: &ArrayView2<f64>) -> Array2<f64> {
    let mut y = Array2::zeros((x.nrows(), 3));
    Zip::from(y.rows_mut())
        .and(x.rows())
        .for_each(|mut yi, xi| {
            yi.assign(&array![g24(&xi), g24_c1(&xi), g24_c2(&xi)]);
        });
    y
}

#[derive(Parser)]
#[command(author, version, about, long_about = None)]
struct Args {
    #[arg(long, default_value = "cobyla")]
    infill_opt: String,
}

fn main() {
    let args = Args::parse();
    
    // Map string argument to InfillOptimizer enum
    let infill_optimizer = match args.infill_opt.as_str() {
        "cobyla" => InfillOptimizer::Cobyla,
        "slsqp" => InfillOptimizer::Slsqp,
        "ipopt" => InfillOptimizer::Ipopt,
        _ => {
            eprintln!("Unknown infill optimizer: {}. Using COBYLA.", args.infill_opt);
            InfillOptimizer::Cobyla
        }
    };

    let xlimits = array![[0., 3.], [0., 4.]];
    let doe = Lhs::new(&xlimits).sample(5);

    let res = EgorBuilder::optimize(f_g24)
        .configure(|config| {
            config
                .n_cstr(2)
                .doe(&doe)
                .max_iters(100)
                .infill_optimizer(infill_optimizer)
                .seed(42)
        })
        .min_within(&xlimits)
        .expect("Egor configured")
        .run()
        .expect("Minimize failure");
    println!(
        "G24 optim result = {} (using {} infill optimizer)",
        res.y_opt, args.infill_opt
    );
}
